example/loop_advection/loop_advection_package.cpp

```diff
@@ -0,0 +1,368 @@
+                        +(cart_dir == 2) ? coords.dx2(j) :
+                                           coords.dx3(k);
+
+        //First term: dEzdy, dExdz, or dEydx
+        const Real dEp2dp1 =(v( cart_dir_p2, b, A,  k, j, i)
+                            -v( cart_dir_p2, b, A,  k - (cart_dir_p1 == 3),
+                                                    j - (cart_dir_p1 == 2),
+                                                    i - (cart_dir_p1 == 1)))/dp1;
+
+        //Second term: dEydz, dEzdx, or dExdy
+        const Real dEp1dp2 =(v( cart_dir_p1, b, A, k, j, i)
+                            -v( cart_dir_p1, b, A, k - (cart_dir_p2 == 3),
+                                                   j - (cart_dir_p2 == 2),
+                                                   i - (cart_dir_p2 == 1)))/dp2;
+
+        v( cart_dir, b, B_face, b, k, j, i) = -(dEp2dp1 - dEp1dp2);
+      });
+  }
+
+}
+
+// Compute v x B to compute E, then take the curl to compute dBdt
+parthenon::TaskStatus calc_dBdt(parthenon::MeshBlock *pmb){
+
+  auto pkg = pmb->packages.Get("loop_advection_package");
+
+  const auto &v1 = pkg->Param<Real>("v1");
+  const auto &v2 = pkg->Param<Real>("v2");
+  const Real v3 = 0;
+
+  auto &v =  SparsePack<B_face,B_cell,E_cell,E_edge>::Make(pmd->meshblock_data.Get());
+
+  using dBdt_face = B_face;
+
+  IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::entire);
+  IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::entire);
+  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::entire);
+
+  //First, compute the cell centered electric field 
+  {
+
+    pmb->par_for(
+      "LoopAdvection::ProblemGenerator:calc_E_cell", 0, vf.dim(5)-1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int k, const int j, const int i) {
+
+        //Average B from the faces to the cell centers
+        v(b, B_cell(0), k, j, i) = 0.5 * (v( 1, b, B_face, k, j, i) + v( 1, b, B_face,  k, j, i+1));
+        v(b, B_cell(1), k, j, i) = 0.5 * (v( 2, b, B_face, k, j, i) + v( 2, b, B_face,  k, j+1, i));
+        v(b, B_cell(2), k, j, i) = 0.5 * (v( 3, b, B_face, k, j, i) + v( 3, b, B_face,  k+1, j, i));
+
+        //Compute E_cell = v X B
+        v(b, E_cell(0), k, j, i) = v2 * B3 - v3 * B2;
+        v(b, E_cell(1), k, j, i) = v3 * B1 - v1 * B3;
+        v(b, E_cell(2), k, j, i) = v1 * B2 - v2 * B1;
+
+      });
+  }
+
+  //Then interpolate E_cell to E_edge
+  //(Do it in one loop for demonstrative purposes)
+  {
+
+    pmb->par_for(
+      "LoopAdvection::ProblemGenerator:calc_E_edge", 0, vf.dim(5)-1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int k, const int j, const int i) {
+        //Average E from the cell centers to cell edges
+        if( k+1 < kb.e && j+1 < jb.e) 
+          v( 1, b, E_edge, k, j, i) = v( 1, b, E_face,   k,   j,   i) + v( 1, b, E_face,   k, j+1,   i);
+                                    + v( 1, b, E_face, k+1,   j,   i) + v( 1, b, E_face, k+1, j+1,   i);
```

Comment:
I am not clear about how the indexing works for faces, edges, and nodes (i.e. for the `i` face does `i -> i-1/2`, `j->j`, and `k->k`? And for the edge `i` does `i -> i`, `j -> j-1/2`, and `k -> k-1/2`). It may be useful to start a doc describing how this all works and also describing what other changes need to be made to the code to include face, edge, and vertex variables (is everything going to be unified in `CellVariable`? What needs to be changed about boundary communication? What changes need to be made to prolongation and restriction? etc.)

Reply:
That indexing is correct, to keep the different arrays for faces/edges all starting at 0.

I'm currently designing around a `FaceVariable` and `EdgeVariable` which keeps separate arrays for each dimension. I first tried working with `CellVariable` but it was awkward.

There are still a lot of changes to prolongation and restriction to be made. There are a few hooks for those in the `loop_advection` example but nothing implemented yet.

Reply:
![image](https://user-images.githubusercontent.com/13839591/181083575-bc1db86e-eefd-4531-9e58-5a4585b044c9.png)

Reply:
Yes, that's the indexing I'm going for. In memory, each edge and face variable would start at 0 for each dimension. Do you see an off-by-one error here that I'm not seeing? I wouldn't be surprised.

Reply:
Based on my diagram, I think I would get 
```c++
v(b, edge_z, E_edge, k, j, i) = 0.25 * (v(b, face_y, E_face k, j, i)  + v(b, face_x, E_face k, j, i) + v(b, face_y, E_face k, j, i-1) + v(b, face_x, E_face k, j-1, i)   )  
```